conn.go

```diff
@@ -806,46 +848,75 @@ func (cn *conn) Query(query string, args []driver.Value) (_ driver.Rows, err err
 	}
 	defer cn.errRecover(&err)
 
+	var r *rows
+
 	// Check to see if we can use the "simpleQuery" interface, which is
 	// *much* faster than going through prepare/exec
 	if len(args) == 0 {
-		return cn.simpleQuery(query)
-	}
-
-	if cn.binaryParameters {
+		r, err = cn.simpleQuery(query)
+	} else if cn.binaryParameters {
 		cn.sendBinaryModeQuery(query, args)
 
 		cn.readParseResponse()
 		cn.readBindResponse()
-		rows := &rows{cn: cn}
-		rows.colNames, rows.colFmts, rows.colTyps = cn.readPortalDescribeResponse()
+		r = &rows{cn: cn}
+		r.colNames, r.colFmts, r.colTyps = cn.readPortalDescribeResponse()
 		cn.postExecuteWorkaround()
-		return rows, nil
 	} else {
 		st := cn.prepareTo(query, "")
 		st.exec(args)
-		return &rows{
+		r = &rows{
 			cn:       cn,
 			colNames: st.colNames,
 			colTyps:  st.colTyps,
 			colFmts:  st.colFmts,
-		}, nil
+		}
+	}
+
+	if r != nil && ctx.Done() != context.Background().Done() {
+		closed := make(chan struct{})
+		go func() {
+			select {
+			case <-ctx.Done():
+				cn.Cancel()
+			case <-closed:
+			}
+		}()
+		r.close = closed
```

Comment:
Maybe consider pulling this whole blob of code out into a function? It is a bit subtle, and it's repeated several times. Maybe something like:

```go
if r != nil && ctx.Done() != nil {
	r.close = watchCancel(ctx, cn.Cancel)
}
```

with 

```go
func watchCancel(ctx context.Context, cancel func()) chan<- struct{} {
	closed := make(chan struct{})
	go func() {
		select {
		case <-ctx.Done():
			cancel()
		case <-closed:
		}
	}()
	return closed
}
```